### ISSUE TITLE
ts_scte_parser.py not decoding SCTE packets

### ISSUE THREAD

**Issue by vasudevavarun** [futzu/threefive#6]
Hello,
I'm trying to decode SCTE packets from a transport stream via  ts_scte_parser.py but it seems like it is not able to decode SCTE packet in the incoming stream while other applications was able to detect it.

Version:  v.2.1.97

I've edited port and mcast_ip as per my requirement

```
    def __init__(self, mcast_ip, if_ip="0.0.0.0", hostname="0.0.0.0", port=5511):
        self.HOST = hostname

if __name__ == "__main__":
    test = TSScte35Parser(mcast_ip="225.5.1.1")
    test.run()
```

**Comment by vladdoster** [comment]
> Hello,
> I'm trying to decode SCTE packets from a transport stream via ts_scte_parser.py but it seems like it is not able to decode SCTE packet in the incoming stream while other applications was able to detect it.
> 
> Version: v.2.1.97
> 
> I've edited port and mcast_ip as per my requirement
> 
> ```
>     def __init__(self, mcast_ip, if_ip="0.0.0.0", hostname="0.0.0.0", port=5511):
>         self.HOST = hostname
> 
> if __name__ == "__main__":
>     test = TSScte35Parser(mcast_ip="225.5.1.1")
>     test.run()
> ```

Could you share either some of the data you're trying to parse or logs? 

What type of exception is it throwing?

**Comment by vladdoster** [comment]
The code you shared doesn't give much context. That isn't the only code you're trying to run right?

**Comment by vasudevavarun** [comment]
Hi,
it is not showing any exception, Just keep listening the stream.

I'm using tsp to analyze the SCTE in the stream and it is able to decode the packet.


```
* pcrextract: PID: 0x1033 (4147), SCTE 35 command SpliceInsert out, exec at PTS 0x095655834
* pcrextract: PID: 0x1033 (4147), SCTE 35 command SpliceInsert in, exec at PTS 0x09598DED4
```

The full code is pasted below, I only pasted the part I've edited.

```
import socket
import threefive
import struct
import io

class TSScte35Parser():
    # TODO: Add unicast / multicast flag or auto-detect based on IP CIDR
    def __init__(self, mcast_ip, if_ip="0.0.0.0", hostname="0.0.0.0", port=5511):
        self.HOST = hostname
        self.PORT = port
        self.MCAST_IP = mcast_ip
        self.IF_IP = if_ip


    def run(self):
        with socket.socket(socket.AF_INET, socket.SOCK_DGRAM, socket.IPPROTO_UDP) as sock:
            self.set_socket_options(sock)
            sock.bind((self.HOST, self.PORT))
            with sock.makefile(mode="b") as socket_file:
                while True:
                    #print(type(socket_file))
                    try:
                        scte = threefive.Stream(socket_file).decode_next()
                        print(scte)
                        if scte:
                            print("Found SCTE-35:", scte)
                            print("--------------------------------")
                    except Exception as e:
                        print("ERROR while decoding TS:", e)
                        pass


    def set_socket_options(self, sock):
        sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        sock.setsockopt(socket.IPPROTO_IP, socket.IP_ADD_MEMBERSHIP, socket.inet_aton(self.MCAST_IP)+socket.inet_aton(self.HOST))
        # TODO: Below is used to specify a specific interface to use for MC subscription

        # group_bin =  socket.inet_aton(self.MCAST_IP)
        # local_bin =  socket.inet_aton(self.IF_IP) # Which interface to IGMP join on

        # sock.setsockopt(socket.IPPROTO_IP, socket.IP_MULTICAST_IF, local_bin)

        # mreq = struct.pack("4sL", group_bin, socket.INADDR_ANY)
        # # mreq = group_bin + socket.inet_aton('6.116.252.42') # Possible IGMPv3 method

        # sock.setsockopt(socket.IPPROTO_IP, socket.IP_ADD_MEMBERSHIP, mreq)

if __name__ == "__main__":
    test = TSScte35Parser(mcast_ip="225.5.1.1")
    test.run()

```

**Comment by futzu** [comment]
Is it a satellite feed with embedded streams?
Can you slice me off a minute or two of the video, and put it somewhere I can get to it?

**Comment by futzu** [comment]
Try this:

```python3
# test.py
import sys
from threefive import Stream

def read_stdin():  
    Stream(sys.stdin.buffer).decode()

if __name__ == "__main__":
    read_stdin()
```

then use tsduck and pipe the stream to it .
```sh
tsp -rtrue -I ip <multicast_ip>:<port_number>  -O - | python3 test.py
```

**Comment by vasudevavarun** [comment]
Hi Adrian,

Following is the output of test.py, The python application kept listening to the stream while tsp decoded it.

![image](https://user-images.githubusercontent.com/16715755/97264858-b4ced600-184b-11eb-84b2-abd7bfaa3a40.png)

I'm going to provide you a sample transport stream dump of what I'm trying to achieve. Can you please share your's email ID.

**Comment by futzu** [comment]
When you run ffprobe on the stream, does it look similar to this?
Does it only shows one stream?
```sh
[mpegts @ 0x55e667178240] Packet corrupt (stream = 0, dts = NOPTS).                                            
[mpegts @ 0x55e667178240] PES packet size mismatch                                                             
[mpegts @ 0x55e667178240] Packet corrupt (stream = 0, dts = NOPTS).                                            
[mpegts @ 0x55e667178240] start time for stream 0 is not set in estimate_timings_from_pts                      
Input #0, mpegts, from '../rus.ts':                                                                            
  Duration: N/A, bitrate: N/A
  Program 800 
    Stream #0:0[0x1000]: Data: bin_data ([6][0][0][0] / 0x0006)
Unsupported codec with id 100359 for input stream 0
```

**Comment by futzu** [comment]
You probably need to demux it. If it has t2Mi embedded streams like this one
https://tsduck.io/streams/?name=russia-t2mi-scte35
```
tsp -rtrue -I ip <multicast_ip>:<port_number>   -p  t2mi --pid 0x1000 -plp 0 | python3 test.py
```
Most likely, you can run the same tsduck command you are running now without the ' -O drop ' at the end and pipe it to  ' python3 test.py.

**Comment by vasudevavarun** [comment]
> You probably need to demux it. If it has t2Mi embedded streams like this one
> https://tsduck.io/streams/?name=russia-t2mi-scte35
> 
> ```
> tsp -rtrue -I ip <multicast_ip>:<port_number>   -p  t2mi --pid 0x1000 -plp 0 | python3 test.py
> ```
> 
> Most likely, you can run the same tsduck command you are running now without the ' -O drop ' at the end and pipe it to ' python3 test.py.

I'm getting same output as tsp while running python3 test.py

> tsp -I ip 225.5.1.1:5511 -l 192.168.37.233 -P pcrextract --scte35 --pid 0x1033 | python3 test.py 
> 
> * pcrextract: PID: 0x1033 (4147), SCTE 35 command SpliceInsert out immediate
> * pcrextract: PID: 0x1033 (4147), SCTE 35 command SpliceInsert in immediate

and ffprobe as you asked

![image](https://user-images.githubusercontent.com/16715755/97399833-f4f98b80-1913-11eb-8053-a6af81b45a6d.png)

**Comment by futzu** [comment]
Can I see the ffmpeg command you ran?

Why do you have two Inputs for tsp ?

Do it without the 
'-P pcrextract --scte35 --pid 0x1033' 

It may be due to partial packets, I'm about to push a new pip build to handle that.

**Comment by vasudevavarun** [comment]
> Do it without the
> '-P pcrextract --scte35 --pid 0x1033'

Still same 
![image](https://user-images.githubusercontent.com/16715755/97493200-9fac9100-198a-11eb-8d41-c1b5471c9592.png)

> Can I see the ffmpeg command you ran?

I'm not running it via ffmpeg, It is from a professional encoder.

> Why do you have two Inputs for tsp ?

I'm not getting this question, Their is only 1 input in tsp.

**Comment by futzu** [comment]
I thought the '-l 192.168.17.233' was another input, what does -l do?
Why is tsp showing scte35 without the ' -P pcrextract --scte35' ?

**Comment by futzu** [comment]
Download a minute or two of the video and upload it to your github account for me so I can figure it out, please.

**Comment by futzu** [comment]
My tsp uses -l to list available inputs and outputs , it doesn't accept any parameters.
Without the -P pcrextract --scte35 , tsp should not be showing scte35 info. 
Are you using a customized tsp or something?
From your ffprobe output, threefive should have no problem parsing it.

**Comment by futzu** [comment]
What happens if you pipe it to mplayer ?

**Comment by vasudevavarun** [comment]
> I thought the '-l 192.168.17.233' was another input, what does -l do?

-l is used when we want listen to the multicast from a particular interface.

>  Why is tsp showing scte35 without the ' -P pcrextract --scte35' ?

I'm running 2 tsp, one is with  -P pcrextract --scte35 just to determine when the scte is coming and below is output of what you instructed to do.

> Download a minute or two of the video and upload it to your github account for me so I can figure it out, please.

Please download it from here
https://drive.google.com/file/d/1Qo0LSsyXMyW5j4RHsc32tati2sqO0RRM/view?usp=sharing

> What happens if you pipe it to mplayer ?

I'm running it in headless mode, pipe to mplayer will be a challenge.

**Comment by futzu** [comment]
Solved as of version 2.2.09. I added support for earlier versions of SCTE-35.
![Screenshot_2020-10-29_18-14-47](https://user-images.githubusercontent.com/52701496/97638281-9a932480-1a12-11eb-91b5-b5c53bc3623a.png)
